ipset.go

```diff
@@ -99,154 +99,135 @@ func (s *IPSet) RemoveSet(b *IPSet) {
-	var points []point
-	for _, r := range s.in {
-		points = append(points, point{r.From, true, true}, point{r.To, true, false})
+	const debug = false
+	if debug {
+		debugf("ranges start in=%v out=%v", s.in, s.out)
 	}
-	for _, r := range s.out {
-		points = append(points, point{r.From, false, true}, point{r.To, false, false})
+	in, ok := mergeIPRanges(s.in)
+	if !ok {
+		return nil
+	}
+	out, ok := mergeIPRanges(s.out)
+	if !ok {
+		return nil
 	}
-	sort.Slice(points, func(i, j int) bool { return points[i].Less(points[j]) })
-	const debug = false
 	if debug {
-		debugf("post-sort:")
-		debugLogPoints(points)
-		debugf("merging...")
+		debugf("ranges sort  in=%v out=%v", in, out)
 	}
 
-	// Now build 'want', like points but with "remove" ranges removed
-	// and adjacent blocks merged, and all elements alternating between
-	// start and end.
-	want := points[:0]
-	var addDepth, removeDepth int
-	for i, p := range points {
-		depth := &addDepth
-		if !p.want {
-			depth = &removeDepth
-		}
-		if p.start {
-			*depth++
-		} else {
-			*depth--
-		}
+	// in and out are sorted in ascending range order, and have no
+	// overlaps within each other. We can run a merge of the two lists
+	// in one pass.
+
+	ret := make([]IPRange, 0, len(in))
+	for len(in) > 0 && len(out) > 0 {
 		if debug {
-			debugf("at[%d] (%+v), add=%v, remove=%v", i, p, addDepth, removeDepth)
-		}
-		if p.start && *depth != 1 {
-			continue
-		}
-		if !p.start && *depth != 0 {
-			continue
+			debugf("step in=%v out=%v", in[0], out[0])
 		}
-		if !p.want && addDepth > 0 {
-			if p.start {
-				// If we're transitioning from a range of
-				// addresses we want to ones we don't, insert
-				// an end marker for the IP before the one we
-				// don't.
-				want = append(want, point{
-					ip:    p.ip.Prior(),
-					want:  true,
-					start: false,
-				})
-			} else {
-				want = append(want, point{
-					ip:    p.ip.Next(),
-					want:  true,
-					start: true,
-				})
+
+		switch {
+		case !out[0].Valid() || !in[0].Valid():
+			// mergeIPRanges should have prevented invalid ranges from
+			// sneaking in.
+			panic("invalid IPRanges during Ranges merge")
+		case out[0].To.Less(in[0].From):
+			// "out" is entirely before "in".
+			//
+			//    out         in
+			// f-------t   f-------t
+			out = out[1:]
+			if debug {
+				debugf("out before in; drop out")
 			}
-		}
-		if !p.want || removeDepth > 0 {
-			continue
-		}
-		// Merge adjacent ranges. Remove prior and skip this
-		// start.
-		if p.start && len(want) > 0 {
-			prior := &want[len(want)-1]
-			if !prior.start && prior.ip == p.ip.Prior() {
-				want = want[:len(want)-1]
-				continue
+		case in[0].To.Less(out[0].From):
+			// "in" is entirely before "out".
+			//
+			//    in         out
+			// f------t   f-------t
+			ret = append(ret, in[0])
+			in = in[1:]
+			if debug {
+				debugf("in before out; append in")
+				debugf("ret=%v", ret)
+			}
+		case out[0].From.lessOrEq(in[0].From) && in[0].To.lessOrEq(out[0].To):
```

Comment:
case foo.supersetOf(bar):

(if you like "superset"?  it's not ambiguous like "covers" which can be partially or fully covered)